README.md
[![PyPI version shields.io](https://img.shields.io/pypi/v/vien.svg)](https://pypi.python.org/pypi/vien/)
[![Generic badge](https://img.shields.io/badge/Python-3.7+-blue.svg)](#)
[![Generic badge](https://img.shields.io/badge/OS-Linux%20|%20macOS-blue.svg)](#)
[![Downloads](https://pepy.tech/badge/vien/month)](https://pepy.tech/project/vien)

# [vien](https://github.com/rtmigo/vien_py#readme)

**VIEN** is a command-line tool for managing
[Python Virtual Environments](https://docs.python.org/3/library/venv.html).

It provides one-line shortcuts for:

- creating and deleting environments
- running commands inside environments
- switching between environments in bash shell

-----

Switching between projects should be simple. Creating environments for the
projects should be simple too.

Ideally it's a short command that I would type even half asleep.

Something like

``` bash
$ vien create 
$ vien shell
```

Not like

``` bash
$ python3 -m venv ./where/to/put/this/.venv
$ source /i/lost/that/.venv/bin/activate
```

<details>
  <summary>Ready-made solutions did not help.</summary><br/>

- [pipenv](https://pipenv.pypa.io/) kind of solved the problem, but brought new
  challenges unrelated to virtual environments
- [virtualenvwrapper](https://virtualenvwrapper.readthedocs.io/) name is easier
  to copy-paste than to type. And its commands are too

</details>

So there is the `vien`. A tool for a half asleep developer.

# Install

`vien` requires Python 3.7+ on Linux or macOS.


<details>
  <summary>Get a working Python ≥3.7, pip3 and venv.</summary><br/>

@ Ubuntu

``` bash
$ sudo apt install -y python3 python3-pip python3-venv
```

@ macOS

``` bash
$ brew install python3
```

Check it works

``` bash
$ python3 --version             # python shows its version
$ python3 -m venv --help        # venv shows help message
$ pip3 install --upgrade pip    # pip upgrades itself
```

----
</details>

Then:

``` bash
$ pip3 install vien
```

Make sure it installed:

``` bash
$ vien      # shows help
```

Upgrade it later:

``` bash
$ pip3 install vien --upgrade
```

# Use

### Example: interactive shell

``` bash
$ cd /path/to/myProject
$ vien create
$ vien shell
```

### Example: running commands

``` bash
$ cd /path/to/myProject
$ vien create
$ vien run pip install --upgrade pip
$ vien run pip install requests lxml
$ vien call main.py
```

# "create" command

`vien create` сreates a virtual environment that will correspond the
**project directory**. Subsequent calls to `vien`
with the same project directory will use the same virtual environment.

``` bash
$ cd /abc/myProject
$ vien create 
```

By default, the current **working directory** is assumed to be the
**project directory**. You can override this with `-p` argument.

``` bash
$ vien -p /abc/myProject create 
```

The `-p` argument works with all commands, not only `create`.

``` bash
$ cd /other/working/dir
$ vien -p /abc/myProject create
$ vien -p /abc/myProject shell
```

### "create": choose the Python version

If you have several versions of Python installed, then virtual environments can
help you switch not only between dependencies, but also between interpreters.

``` bash
$ cd /abc/my_old_project
$ vien run python --version  # 3.7  

$ cd /abc/my_newer_project
$ vien run python --version  # 3.9 
```

A project directory can be bound to only one specific interpreter.

To indicate which interpreter to use with the project, provide an argument
pointing the executable.

``` bash
$ vien create /usr/local/opt/python@3.8/bin/python3
```

In many cases, a shorter command will also work. If the needed interpreter can
be executed in the shell as `python3.8`, you can try

``` bash
$ vien create python3.8
```

When `create` is called with no argument, `vien` will use the Python interpreter
that is running `vien` itself. For example, if you used Python 3.9
to `pip install vien`, then it is the Python 3.9 runs `vien`, and this Python
3.9 will be used in the virtual environment.

# "shell" command

`vien shell` starts interactive bash session in the virtual environment.

``` bash	
$ cd /path/to/myProject
$ vien shell

(myProject)$ _
```

Now you are inside the virtual environment.

``` bash	
(myProject)$ which python3             # now we are using separate copy of Python
(myProject)$ echo $PATH                # everything is slightly different

(myProject)$ pip3 install requests     # installs packages into virtual environment
(myProject)$ python3 use_requests.py   # runs inside the virtual environment
```

Get out of the virtual environment:

``` bash
(myProject)$ exit

$ _
```

Now you're back.

With shell pipes, you can specify what the shell should execute right in the
command line.

``` bash
$ echo 'which python3 && echo $PATH' | vien shell
```

# "run" command

`vien run COMMAND` runs a shell command in the virtual environment.

``` bash 		
$ cd /path/to/myProject
$ vien run python3 use_requests.py arg1 arg2  # runs script in virtual environment
$ vien run pip3 install requests              # installs packages into virtual environment
```

<details>
  <summary>is an equivalent to</summary><br/>

``` bash 		
$ cd /path/to/myProject

$ source /path/to/the/venv/bin/activate
$ python3 use_requests.py arg1 arg2
$ /path/to/the/venv/bin/deactivate

$ source /path/to/the/venv/bin/activate
$ pip3 install requests
$ /path/to/the/venv/bin/deactivate
```

</details>

call                            | run
--------------------------------|-----------------------------------------------
Runs only `python file.py` or `python -m module` | Can run any shell command: `pip3`, `cd`, etc.
Starts one python process       | Starts two processes: parent shell and child python

# "call" command

`vien call PYFILE` executes a `.py` script in the virtual environment.

### "call": running file as a file

``` bash
$ cd /abc/myProject
$ vien call pkg/module.py
  
# runs [python pkg/module.py]
```

### "call": running file as a module

If the `.py` file name is preceded by the `-m` argument, we will run it with
`python -m MODULE`. Running in this manner often simplifies importing other
modules from the program.

``` bash
$ cd /abc/myProject
$ vien call -m /abc/myProject/pkg/sub/module.py

# runs [python -m pkg.sub.module]
# project dir: /abc/myProject
# working dir: /abc/myProject
```

- `module.py` must be located somewhere inside the `/abc/myProject`
- parent subdirectories such as `pkg` and `sub` must be importable, i.e. must
  contain
  `__init__.py`
- the project directory will be inserted into `$PYTHONPATH`, making
  `pkg.sub.module` resolvable from `/abc/myProject` to a file

The project directory can be specified not only by the working directory, but
also by the `-p` argument.

The `call` command only accepts `.py` files, no module names.

``` bash
# ERROR: there is no file named pkg.module
$ vien call -m pkg.module 
```

### "call": passing arguments to Python and to the program

Arguments following the `call` command are passed to the python executable.

``` bash 		
$ vien call -B -OO -m package/main.py arg1 arg2  

# runs [python -B -OO -m package.main arg1 arg2]
```

### "call": project directory

The optional `-p` argument can be specified before the `call` word. It allows
you to set the project directory **relative** to the parent directory of the
**file** being run.

``` bash
$ cd /far/away
$ vien -p /abc/myProject call -m /abc/myProject/pkg/sub/module.py

# runs [python -m pkg.sub.module]
# project dir: /abc/myProject
# working dir: /far/away
```

``` bash
$ cd /far/away
$ vien -p ../.. call -m /abc/myProject/pkg/sub/module.py

# runs [python -m pkg.sub.module]
# project dir: /abc/myProject  (/abc/myProject/pkg/sub/../..)
# working dir: /far/away
```

``` bash
$ cd /abc/myProject/pkg
$ vien -p ../.. call -m sub/module.py

# runs [python -m pkg.sub.module]
# project dir: /abc/myProject  (/abc/myProject/pkg/sub/../..)
# working dir: /abc/myProject/pkg
```

# "delete" command

`vien delete` deletes the virtual environment.

``` bash
$ cd /path/to/myProject
$ vien delete 
```

# "recreate" command

`vien recreate` old and creates new virtual environment.

If you decided to start from scratch:

``` bash
$ cd /path/to/myProject
$ vien recreate 
```

If you decided to change the Python version:

``` bash
$ cd /path/to/myProject
$ vien recreate /usr/local/opt/python@3.10/bin/python3
```

# --project-dir, -p

This option must appear after `vien`, but before the command.

``` bash
vien -p /abc/myProject create ...
vien -p /abc/myProject run ...
vien -p /abc/myProject shell ...
```

If `--project-dir` is specified, it is the project directory.

If `--project-dir` is not specified, then all commands assume that the current
working directory is the project directory.

The next two calls use the same project directory and the same virtual
environment. However, the working directory is different.

``` bash
cd /abc/myProject
vien run python3 /abc/myProject/main.py
```

``` bash
cd /any/where
vien -p /abc/myProject run python3 /abc/myProject/main.py
```

If `--project-dir` is specified as a **relative path**, its interpretation
depends on the command.

- For the `call` command, this is a path relative to the parent directory of
  the `.py` file being run
- For other commands, this is a path relative to the current working directory

# Virtual environments location

By default, `vien` places virtual environments in the `$HOME/.vien` directory.

|project dir|virtual environment dir|
|-----|----|
|`/abc/thisProject`|`$HOME/.vien/thisProject_venv`|
|`/abc/otherProject`|`$HOME/.vien/otherProject_venv`|
|`/moved/to/otherProject`|`$HOME/.vien/otherProject_venv`|

Only the local name of the project directory matters.

If you're not happy with the default, you can set the environment
variable `VIENDIR`:

``` bash
$ export VIENDIR="/x/y/z"
```

So for the project `aaa` the virtual environment will be located
in `/x/y/z/aaa_venv`.

The `_venv` suffix tells the utility that this directory can be safely removed.

# Shebang

On POSIX systems, you can make a `.py` file executable, with `vien` executing it
inside a virtual environment.

Insert the shebang line to the top of the file you want to run. The value of the
shebang depends on the location of the file relative to the project directory.

File                            | Shebang line
--------------------------------|--------------------------------------
`myProject/runme.py`            | `#!/usr/bin/env vien -p . call -m`
`myProject/pkg/runme.py`        | `#!/usr/bin/env vien -p .. call -m`
`myProject/pkg/subpkg/runme.py` | `#!/usr/bin/env vien -p ../.. call -m`

After inserting the shebang, make the file executable:

``` bash
$ chmod +x runme.py  
```

Now you can run the `runme.py` directly from command line. This will use the
virtual environment associated with the `myProject`. The working directory can
be anything.

``` bash
# runs the runme.py in virtual environment for myProject

$ cd anywhere/somewhere
$ /abc/myProject/pkg/main.py   
```

# Shell prompt

By default the `vien shell` adds a prefix to
the [`$PS1`](https://wiki.archlinux.org/index.php/Bash/Prompt_customization)
bash prompt.

``` bash
user@host$ cd /abc/myProject
user@host$ vien shell

(myProject)user@host$ _
```

So you can see, which virtual environment you're using.

If you customized your `PS1`, it may not work as expected.

``` bash
personalized:prompt> cd /abc/myProject
personalized:prompt> vien shell

(myProject)user@host$ _
```

It can be fixed by providing `PS1` variable to `vien` like that:

``` bash
personalized:prompt> cd /abc/myProject
personalized:prompt> PS1=$PS1 vien shell

(myProject)personalized:prompt> _
```

To avoid doing this each time, `export` your `PS1` to make it available for
subprocesses.

CHANGELOG.md
# 8.1.3

- fixed `PATH` issue on Linux

# 8.1.1

- fixed shell prompt color bug on Linux

# 8.1

- The `create` and `recreate` commands show more detailed information about the
  created virtual environment

# 8.0

- The `call` command now supports the `-m` option, that runs the file as a
  module.

# 7.1

- Fixed: `call` command sometimes received incorrect `$PYTHONPATH` values on systems
  with multiple versions of Python

# 7.0

- The `create`, `call`, `delete`, `recreate` and `path` commands work on Windows.

# 6.0

- The `create` and `recreate` commands without arguments will create a virtual
  environment with the same interpreter that started `vien`, instead of the
  interpreter available as `python3`
- Fixed: virtual environments were not cleared (`venv --clear`) before deletion  
- Attempting to run the program on Windows will display a message stating that
  the system is not fully supported

# 5.0

- Option `--project-dir` aka `-p` is now supported by all commands
- The `vien call -p` format (with the `-p` option after the command) is deprecated but still works

# 4.4

- `call` is now faster as it launches Python directly without spawning an extra
  shell process

# 4.3

- A much shorter help message now appears when running `vien` without parameters
- Fixed: `call` printed debug message to stdout

# 4.2

- Trying to `call` a non-existent file now prints a short error message instead of an exception
- A message about a nonexistent environment displays a hint on how to create it  

# 4.1

- Fixed: `$VIENDIR` did not change the storage directory of the virtual
  environments

# 4.0

- "call" command added
- New versioning scheme
- Fixed: the 'run' command could be run when the virtual environment does not
  exist

# 0.3.7

- Fixed: "run" and "shell" commands returned exit code 0 regardless of the exit
  code of the child process 

vien/main.py
# SPDX-FileCopyrightText: (c) 2020 Artëm IG <github.com/rtmigo>
# SPDX-License-Identifier: BSD-3-Clause

from __future__ import annotations

import json
import os
import shutil
import subprocess
import sys
import shlex
from pathlib import Path
from typing import *

from vien import is_posix
from vien._common import need_posix, is_windows, need_windows
from vien.arg_parser import Commands, Parsed
from vien.bash_runner import run_as_bash_script
from vien._call_funcs import relative_fn_to_module_name, relative_inner_path
from vien.call_parser import ParsedCall, list_left_partition
from vien.colors import Colors
from vien.escaping_cmd import cmd_escape_arg
from vien.exceptions import ChildExit, VenvExistsExit, VenvDoesNotExistExit, \
    PyFileNotFoundExit, PyFileArgNotFoundExit, FailedToCreateVenvExit, \
    FailedToClearVenvExit, CannotFindExecutableExit

verbose = False


def exe_name() -> str:
    return os.path.basename(sys.argv[0])


def get_vien_dir() -> Path:
    path_from_var = os.environ.get("VIENDIR")
    if path_from_var:
        path_from_var = os.path.expandvars(path_from_var)
        path_from_var = os.path.expanduser(path_from_var)
        return Path(path_from_var)
    else:
        # It looks like storing dot files in the home directory
        # is the de facto standard for both worlds.
        #
        # App    | POSIX     | Windows
        # -------|-----------|----------------------
        # VSCode | ~/.vscode | %USERPROFILE%\.vscode
        # AWS:   | ~/.aws    | %USERPROFILE%\.aws
        #
        return Path.home() / ".vien"


def run_bash_sequence(commands: List[str], env: Optional[Dict] = None) -> int:
    need_posix()

    # command || exit /b 666

    lines = [
        # shebang not necessary as we specify executable in subprocess.call
        "set -e",  # fail on first error
    ]

    lines.extend(commands)

    # Ubuntu really needs executable='/bin/bash'.
    # Otherwise the command is executed in /bin/sh, ignoring the hashbang,
    # but SH fails to execute commands like 'source'

    return subprocess.call("\n".join(lines),
                           shell=True,
                           executable='/bin/bash',
                           env=env)


def run_cmdexe_sequence(commands: List[str], env: Optional[Dict] = None) -> int:
    # todo test independently
    # This function does not work "officially" yet.

    need_windows()

    # raise NotImplemented

    # https://stackoverflow.com/questions/734598/how-do-i-make-a-batch-file-terminate-upon-encountering-an-error

    # unlike bash, cmd.exe returns exit code zero even if last command returned
    # non-zero. There is also no evident way to turn on 'set -e' mode, i.e.
    # exit on the first failure.
    #
    # We'll just glue all the commands with &&

    glued = " && ".join(f'( {c} )' for c in commands)

    # print(f"CMD running {glued}")

    return subprocess.call(glued,
                           shell=True,
                           # executable='/bin/bash',
                           env=env)


def venv_dir_to_python_exe(venv_dir: Path) -> Path:
    # this method is being tested indirectly each time the venv is created:
    # vien prints the path to executable after running this function

    if is_posix:
        parent = venv_dir / "bin"
        basenames = "python", "python3"
    else:
        parent = venv_dir / "Scripts"
        basenames = "python.exe", "python3.exe"

    for name in basenames:
        executable = parent / name
        if executable.exists():
            return executable

    raise Exception(f"Cannot find the Python interpreter in {venv_dir}.")


def windows_cmdexe_activate(venv_dir: Path) -> Path:
    # https://docs.python.org/3/library/venv.html
    assert is_windows
    return venv_dir / 'Scripts' / 'activate.bat'


def posix_bash_activate(venv_dir: Path) -> Path:
    # https://docs.python.org/3/library/venv.html
    assert is_posix
    return venv_dir / 'bin' / 'activate'


def arg_to_python_interpreter(argument: Optional[str]) -> str:
    if argument is None:
        return sys.executable
    exe = shutil.which(argument)
    if not exe:
        raise CannotFindExecutableExit(argument)
    return exe


def main_create(venv_dir: Path, interpreter: Optional[str]):
    if venv_dir.exists():
        raise VenvExistsExit("Virtualenv already exists.")

    exe = arg_to_python_interpreter(interpreter)

    print(f"Creating {venv_dir}")

    result = subprocess.run([exe, "-m", "venv", str(venv_dir)])
    if result.returncode == 0:
        print()
        print("The Python executable:")
        print(str(venv_dir_to_python_exe(venv_dir)))
    else:
        raise FailedToCreateVenvExit(venv_dir)


def main_delete(venv_dir: Path):
    if "_venv" not in venv_dir.name:
        raise ValueError(venv_dir)
    if not venv_dir.exists():
        raise VenvDoesNotExistExit(venv_dir)

    # todo try to use the same executable that created the environment
    # If we use sys.executable, we may clear the venv in some incompatible way.
    # But we can't just use executable from the venv when clearing it:
    # Windows will fail with [WinError 5] Access is denied: '...python.exe'

    # todo check we are not running the same executable we about to delete
    # python_exe = venv_dir_to_python_exe(venv_dir)
    print(f"Clearing {venv_dir}")

    result = subprocess.run(
        [sys.executable, "-m", "venv", "--clear", str(venv_dir)],
        capture_output=True, encoding=sys.stdout.encoding)
    if result.returncode != 0:
        print(f"stdout: {result.stdout}")
        print(f"stderr: {result.stderr}")
        raise FailedToClearVenvExit(venv_dir)
    print(f"Deleting {venv_dir}")
    shutil.rmtree(str(venv_dir))


def main_recreate(venv_dir: Path, interpreter: Optional[str]):
    if venv_dir.exists():
        main_delete(venv_dir)
    main_create(venv_dir=venv_dir, interpreter=interpreter)


def guess_bash_ps1():
    """Returns the default BASH prompt."""

    # TL;DR PS1 is often inaccessible for child processes of BASH. It means,
    # for scripts too.
    #
    # AG 2021: PS1 is not an environment variable, but a local variable of
    # the shell [2019](https://stackoverflow.com/a/54999265). It seems to be
    # true for both MacOS 10.13 and Ubuntu 18.04.
    #
    # We can see PS1 by typing "echo $PS1" to the prompt, but ...
    #
    # 1) script.sh with `echo $PS1`        | prints nothing MacOS & Ubuntu
    #
    # 2) module.py with                    | prints Null MacOS & Ubuntu
    #    `print(os.environ.get("PS1")      |
    #
    # 3) `bash -i -c "echo $PS1"`          | seems to be OK in Ubuntu
    #     from command line                |
    #
    # 4) `zsh -i -c "echo $PS1"`           | looks like a normal prompt in OSX
    #     from command line                |
    #
    # In Ubuntu (3) returns the same prompt that in used by terminal by default.
    # Although if the user customized their PS1, no guarantees, that (3) will
    # return the updated value.
    #
    # For MacOS, the prompt printed by (3) in not the same as seen in terminal
    # app. It returns boring "bash-3.2" instead of expected "host:dir user$".
    #
    # (4) on MacOS seems to return the correct "host:dir user$", but it is in
    # ZSH format.

    # try to return $PS1 environment variable:
    env_var = os.environ.get("PS1")
    if env_var is not None:
        return env_var

    # for MacOS return predefined constant PS1
    import platform
    if platform.system() == "Darwin":
        return r"\h:\W \u\$"  # default for MacOS up to Catalina

    # hope for the best in other systems
    return subprocess.check_output(
        ['/bin/bash', '-i', '-c', 'echo $PS1']).decode().rstrip()


def main_shell(dirs: Dirs, input: Optional[str], input_delay: Optional[float]):
    dirs.venv_must_exist()

    activate_path_quoted = shlex.quote(
        str(dirs.venv_dir / "bin" / "activate"))

    old_ps1 = os.environ.get("PS1") or guess_bash_ps1()

    if not old_ps1:
        old_ps1 = r"\h:\W \u\$"  # default from MacOS

    color_start = Colors.YELLOW
    color_end = Colors.NOCOLOR

    venv_name = dirs.project_dir.name
    new_ps1 = f"{color_start}({venv_name}){color_end}:{old_ps1} "

    commands = [f'source {activate_path_quoted}']

    bashrc_file = Path(os.path.expanduser("~/.bashrc"))

    if bashrc_file.exists():
        # Ubuntu
        commands.append(
            f"exec bash --rcfile <(cat {json.dumps(str(bashrc_file))} "
            f"&& echo 'PS1={json.dumps(new_ps1)}')")
    else:
        # MacOS
        commands.append(f"PS1={json.dumps(new_ps1)} exec bash")

    # we will use [input] for testing: we will send a command to the stdin of
    # the interactive sub-shell and later check whether the command was
    # executed.
    #
    # We will also provide [input_delay] parameter. This allows the check
    # whether
    # the sub-shell was really interactive: did it wait for the input
    #
    # Surprisingly, the sub-shell will immediately close after executing the
    # command.  It seems it closes immediately after the subprocess.Popen
    # closes the stdin. So it will not wait for "exit". But it serves the
    # task well

    cp = run_as_bash_script("\n".join(commands),
                            input=input.encode() if input else None,
                            input_delay=input_delay,
                            env=child_env(dirs.project_dir))

    # the vien will return the same exit code as the shell returned
    raise ChildExit(cp.returncode)


def bash_args_to_str(args: List[str]) -> str:
    return ' '.join(shlex.quote(arg) for arg in args)


def cmdexe_args_to_str(args: List[str]) -> str:
    return ' '.join(cmd_escape_arg(arg) for arg in args)


def main_run(dirs: Dirs, command: List[str]):
    dirs.venv_must_exist()

    sequence: List[str] = list()

    if is_posix:
        activate_file = posix_bash_activate(dirs.venv_dir)
        sequence.append(f'source {shlex.quote(str(activate_file))}')
        sequence.append(bash_args_to_str(command))
        run_func = run_bash_sequence
    elif is_windows:
        activate_file = windows_cmdexe_activate(dirs.venv_dir)
        sequence.append(f'CALL {activate_file}"')
        sequence.append(cmdexe_args_to_str(command))
        run_func = run_cmdexe_sequence
    else:
        raise AssertionError("Unexpected OS")
    #    activate_file = (dirs.venv_dir / 'bin' / 'activate').absolute()
    if not activate_file.exists():
        raise FileNotFoundError(activate_file)

    exit_code = run_func(sequence, env=child_env(dirs.project_dir))
    raise ChildExit(exit_code)


class Dirs:
    def __init__(self, project_dir: Union[str, Path] = '.'):
        self.project_dir = Path(project_dir).absolute()
        self.venv_dir = get_vien_dir() / (self.project_dir.name + "_venv")
        if verbose:
            print(f"Proj dir: {self.project_dir}")
            print(f"Venv dir: {self.venv_dir}")

    def venv_must_exist(self) -> Dirs:
        if not self.venv_dir.exists():
            raise VenvDoesNotExistExit(self.venv_dir)
        return self


def _insert_into_pythonpath(insert_me: str) -> str:
    # https://docs.python.org/3/using/cmdline.html#envvar-PYTHONPATH
    # "The format is the same as the shell’s PATH: one or more directory
    # pathnames separated by os.pathsep (e.g. colons on Unix or semicolons
    # on Windows)"
    #
    ####
    #
    # $PYTHONPATH and sys.path are not the same.
    # - $PYTHONPATH is the variable to be used by all Python interpreters
    #   in the current shell environment
    # - sys.path is values for the current interpreter
    #
    # Passing values from own sys.path to other interpreter would be a mistake.
    # The current interpreter may be Python 3.7, and the other interpreter is
    # Python 3.9. Our sys.path leads to 3.7 system libraries, that are
    # unnecessary and even cause errors if added to $PYTHONPATH of 3.9.
    #
    ######
    # keeping it simple: not parsing the old string, just adding a prefix

    old = os.environ.get("PYTHONPATH", '')
    return f'{insert_me}{os.pathsep}{old}'


def child_env(proj_path: Path) -> Optional[Dict]:
    env: Optional[Dict]
    if proj_path != Path.cwd():
        return {
            **os.environ,
            'PYTHONPATH': _insert_into_pythonpath(str(proj_path))
        }
    else:
        return None


def replace_arg(args: List[str], old: str, new: List[str]) -> List[str]:
    """Replaces first occurrence of `old` with a list of `new` items (zero or
    more items). Raises exception if `old` not found.
    """

    # todo remove?
    result = list()
    replaced = False
    for arg in args:
        if not replaced and arg == old:
            result.extend(new)
            replaced = True
        else:
            result.append(arg)

    assert replaced
    return result


def main_call(parsed: Parsed, dirs: Dirs):
    dirs.venv_must_exist()

    parsed_call = ParsedCall(parsed.args)
    assert parsed_call.file is not None

    if not os.path.exists(parsed_call.file):
        raise PyFileNotFoundExit(Path(parsed_call.file))

    if parsed_call.before_filename == "-m":
        # todo unit test
        # /abc/project/package/module.py -> package/module.py
        relative = relative_inner_path(parsed_call.file, dirs.project_dir)
        # package/module.py -> package.module
        module_name = relative_fn_to_module_name(relative)
        # replacing the filename in args with the module name.
        # It is already prefixed with -m
        args = parsed_call.args.copy()
        args[parsed_call.file_idx] = module_name
        # args to python are those after 'call' word
        _, args_to_python = list_left_partition(args, 'call')
        assert '-m' in args_to_python
        assert module_name in args_to_python
    else:
        args_to_python = parsed.args_to_python

    python_exe = venv_dir_to_python_exe(dirs.venv_dir)

    assert len(args_to_python) > 0
    args = [str(python_exe)] + args_to_python

    cp = subprocess.run(args, env=child_env(dirs.project_dir))

    raise ChildExit(cp.returncode)


def normalize_path(reference: Path, path: Path) -> Path:
    # todo test
    if path.is_absolute():
        return path
    return Path(os.path.normpath(reference / path))


def get_project_dir(parsed: Parsed) -> Path:
    if parsed.project_dir_arg is not None:
        if parsed.command == Commands.call:
            # for the 'call' the reference dir is the parent or .py file
            pyfile = ParsedCall(parsed.args).file
            if pyfile is None:
                raise PyFileArgNotFoundExit
            reference_dir = Path(pyfile).parent.absolute()
        else:
            # for other commands the reference dir is cwd
            reference_dir = Path(".").absolute()
        project_dir = normalize_path(reference_dir,
                                     Path(parsed.project_dir_arg))
    else:
        # when the --project-dir is not specified, the project dir is cwd
        assert parsed.project_dir_arg is None
        project_dir = Path('.')
    project_dir = project_dir.absolute()
    return project_dir


def main_entry_point(args: Optional[List[str]] = None):
    parsed = Parsed(args)

    dirs = Dirs(project_dir=get_project_dir(parsed))

    if parsed.command == Commands.create:
        main_create(dirs.venv_dir, parsed.python_executable)
    elif parsed.command == Commands.recreate:
        main_recreate(dirs.venv_dir,
                      parsed.python_executable)  # todo .existing()?
    elif parsed.command == Commands.delete:  # todo move 'existing' check from func?
        main_delete(dirs.venv_dir)
    elif parsed.command == Commands.path:
        print(dirs.venv_dir)  # does not need to be existing
    elif parsed.command == Commands.run:
        # todo allow running commands from strings
        main_run(dirs.venv_must_exist(), parsed.run_args)
    elif parsed.command == Commands.call:

        main_call(parsed, dirs)

    elif parsed.command == Commands.shell:
        main_shell(dirs, parsed.shell_input, parsed.shell_delay)
    else:
        raise ValueError
